src/servers/src/prom_store.rs

```diff
@@ -300,6 +301,59 @@ fn recordbatch_to_timeseries(table: &str, recordbatch: RecordBatch) -> Result<Ve
     Ok(timeseries_map.into_values().collect())
 }
 
+pub fn to_grpc_row_insert_requests(request: WriteRequest) -> Result<(RowInsertRequests, usize)> {
+    let mut multi_table_data = MultiTableData::new();
+
+    for series in &request.timeseries {
+        let table_name = &series
+            .labels
+            .iter()
+            .find(|label| {
+                // The metric name is a special label
+                label.name == METRIC_NAME_LABEL
+            })
+            .context(error::InvalidPromRemoteRequestSnafu {
+                msg: "missing '__name__' label in time-series",
+            })?
+            .value;
+
+        // The metric name is a special label,
+        // num_columns = labels.len() - 1 + 1 (value) + 1 (timestamp)
+        let num_columns = series.labels.len() + 1;
+
+        let table_data = multi_table_data.get_or_default_table_data(
+            table_name,
+            num_columns,
+            series.samples.len(),
+        );
+
+        for Sample { value, timestamp } in &series.samples {
+            let mut one_row = table_data.alloc_one_row();
+
+            // labels
+            let kvs = series
+                .labels
+                .iter()
+                .skip_while(|label| label.name == METRIC_NAME_LABEL)
+                .map(|label| (label.name.as_str(), label.value.as_str()));
```

Comment:
issue: we should use filter_map. [skip_while()](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.skip_while) always yield the element once the predicate returns false
```suggestion
                .filter_map(|label| if label.name == METRIC_NAME_LABEL { None } else { Some((label.name.as_str(), label.value.as_str()) });
```

The following code outputs `[1, -1, 0, 1]`
```rust
fn main() {
let a = [1, -1i32, 0, 1];
let b: Vec<_> = a.iter().skip_while(|x| x.is_negative()).collect();
println!("{b:?}");
}
```